interfaces/builtin/shared_memory.go

```diff
@@ -0,0 +1,236 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+const sharedMemorySummary = `allows two snaps to use predefined shared memory objects`
+
+const sharedMemoryBaseDeclarationPlugs = `
+  shared-memory:
+    allow-installation: true
+    allow-connection:
+      slot-attributes:
+        shared-memory: $PLUG(shared-memory)
+    allow-auto-connection:
+      slot-publisher-id:
+        - $PLUG_PUBLISHER_ID
+      slot-attributes:
+        shared-memory: $PLUG(shared-memory)
+`
+
+const sharedMemoryBaseDeclarationSlots = `
+  shared-memory:
+    allow-installation: false
+    deny-connection: true
+    deny-auto-connection: true
+`
+
+func validateSharedMemoryPath(path string) error {
+	if len(path) == 0 {
+		return fmt.Errorf("shared-memory interface path is empty")
+	}
+
+	if strings.TrimSpace(path) != path {
+		return fmt.Errorf("shared-memory interface path has leading or trailing spaces: %q", path)
+	}
+
+	// TODO: allow "*" as a globbing character; figure out if more AARE should be allowed
+	if err := apparmor.ValidateNoAppArmorRegexp(path); err != nil {
+		return fmt.Errorf("shared-memory interface path is invalid: %v", err)
+	}
+
+	// TODO: consider whether we should remove this check and allow full SHM path
+	if strings.Contains(path, "/") {
+		return fmt.Errorf("shared-memory interface path should not contain '/': %q", path)
+	}
+
+	// The check above protects from most unclean paths, but one could still specify ".."
+	if !cleanSubPath(path) {
+		return fmt.Errorf("shared-memory interface path is not clean: %q", path)
+	}
+
+	return nil
+}
+
+func readStringListAttribute(attrer interfaces.Attrer, key string) ([]string, error) {
```

Comment:
nitpick: read is unusual here as a prefix,  it seems we tend to use nothing or "get" or "verify". Also here it might seem to refer to the "read" attribute but is actually for both read and write. My preference is to just call this:  "stringListrAttribute".

Reply:
Done!